src/net.cpp

```diff
@@ -738,6 +864,58 @@ void V1TransportSerializer::prepareForTransport(CSerializedNetMsg& msg, std::vec
     CVectorWriter{SER_NETWORK, INIT_PROTO_VERSION, header, 0, hdr};
 }
 
+void V2TransportSerializer::prepareForTransport(CSerializedNetMsg& msg, std::vector<unsigned char>& header)
+{
+    size_t serialized_command_size = 1; // short-IDs are 1 byte
+    const uint8_t cmd_short_id = GetShortCommandIDFromCommand(msg.command);
+    if (cmd_short_id == 0) {
+        // message command without an assigned short-ID
+        assert(msg.command.size() <= NET_P2P_V2_CMD_MAX_CHARS_SIZE);
+        // encode as varstr, max 12 chars
+        serialized_command_size = ::GetSerializeSize(msg.command, PROTOCOL_VERSION);
+    }
+    // prepare the packet length that will later be encrypted and part of the MAC (AD)
+    // the packet length excludes the 16 byte MAC tag
+    uint32_t packet_length = serialized_command_size + msg.data.size();
+
+    // prepare the packet length & message command and reserve 4 bytes (3bytes AD + 1byte short-ID)
+    std::vector<unsigned char> serialized_header(CHACHA20_POLY1305_AEAD_AAD_LEN + 1);
```

Comment:
I think this should be `header.reserve(CHACHA20_POLY1305_AEAD_AAD_LEN + 1);` and all instances of `serialized_header` should be replaced with `header`.

Reply:
Note, after I changed this, the V2 tests fail:

<details><summary>log</summary>
<p>

```
neha@mumford:~/src/bitcoin (HEAD detached at e13accd43a)$ src/test/test_bitcoin -t net_tests -l test_suite
Running 11 test cases...
Entering test module "Bitcoin Core Test Suite"
test/net_tests.cpp(83): Entering test suite "net_tests"
test/net_tests.cpp(85): Entering test case "cnode_listen_port"
test/net_tests.cpp(85): Leaving test case "cnode_listen_port"; testing time: 17203us
test/net_tests.cpp(97): Entering test case "caddrdb_read"
test/net_tests.cpp(97): Leaving test case "caddrdb_read"; testing time: 3942us
test/net_tests.cpp(143): Entering test case "caddrdb_read_corrupted"
test/net_tests.cpp(143): Leaving test case "caddrdb_read_corrupted"; testing time: 3158us
test/net_tests.cpp(173): Entering test case "cnode_simple_test"
test/net_tests.cpp(173): Leaving test case "cnode_simple_test"; testing time: 3249us
test/net_tests.cpp(198): Entering test case "ipv4_peer_with_ipv6_addrMe_test"
test/net_tests.cpp(198): Leaving test case "ipv4_peer_with_ipv6_addrMe_test"; testing time: 2877us
test/net_tests.cpp(236): Entering test case "LimitedAndReachable_Network"
test/net_tests.cpp(236): Leaving test case "LimitedAndReachable_Network"; testing time: 2849us
test/net_tests.cpp(259): Entering test case "LimitedAndReachable_NetworkCaseUnroutableAndInternal"
test/net_tests.cpp(259): Leaving test case "LimitedAndReachable_NetworkCaseUnroutableAndInternal"; testing time: 2807us
test/net_tests.cpp(282): Entering test case "LimitedAndReachable_CNetAddr"
test/net_tests.cpp(282): Leaving test case "LimitedAndReachable_CNetAddr"; testing time: 2843us
test/net_tests.cpp(296): Entering test case "LocalAddress_BasicLifecycle"
test/net_tests.cpp(296): Leaving test case "LocalAddress_BasicLifecycle"; testing time: 2845us
test/net_tests.cpp(310): Entering test case "PoissonNextSend"
test/net_tests.cpp(310): Leaving test case "PoissonNextSend"; testing time: 2813us
test/net_tests.cpp(371): Entering test case "net_v2"
test/net_tests.cpp(361): error: in "net_tests/net_v2": check deserializer->Complete() has failed
test/net_tests.cpp(362): error: in "net_tests/net_v2": check read_bytes == msg.data.size() + serialized_header.size() has failed [23 != 24]
test_bitcoin: net.cpp:783: virtual CNetMessage V2TransportDeserializer::GetMessage(const unsigned char (&)[4], int64_t): Assertion `Complete()' failed.
unknown location(0): fatal error: in "net_tests/net_v2": signal: SIGABRT (application abort requested)
test/net_tests.cpp(362): last checkpoint
test/net_tests.cpp(371): Leaving test case "net_v2"; testing time: 3128us
test/net_tests.cpp(83): Leaving test suite "net_tests"; testing time: 47830us
Leaving test module "Bitcoin Core Test Suite"; testing time: 47866us

*** 3 failures are detected in the test module "Bitcoin Core Test Suite"
```
</details>

Reply:
I see your point.
The pass-by-reference `header` field in `prepareForTransport()` is only used in V1.
In V2, there is actually no header (well, you could argue that the MAC-tag and the encrypted length is the header but since the MAC is at the end of the message, it makes little sense to try to use that header field).

Therefor, in V2, the pass-by-reference `header` field is unused and the whole message (including the AAD & MAC) is part of the message data/payload (which I think is conceptual fine).